### PR TITLE
refactor(core): All calls to supplyData should use a distinct context type (no-changelog)

### PR DESCRIPTION
**Pull request by netroy**
## Summary

When we added nodes that have a `.supplyData`, we did not create a distinct execution context type for these calls, and reused `IExecuteFunctions` for those calls. Not only is this semantically incorrect, it also created issues that we fixed by monkey-patching the execution context of the parent node.
This PR creates a distinct `ISupplyDataFunctions` interface for the execution context for these calls.

PS: This PR duplicates some code in `NodeExecutionFunctions`. I've created CAT-271 to tackle this and to add tests for these context objects. 

## Related Linear tickets, Github issues, and Community forum posts

CAT-269

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

File: packages/core/src/NodeExecuteFunctions.ts [hunk 1]
```diff
@@ -4059,6 +4008,258 @@ export function getExecuteFunctions(
...
+			),
+		continueOnFail: () => continueOnFail(node),
+		evaluateExpression: (expression: string, itemIndex: number) =>
+			workflow.expression.resolveSimpleParameterValue(
+				`=${expression}`,
+				{},
+				runExecutionData,
+				runIndex,
+				itemIndex,
+				// TODO: revert this back to `node.name` when we stop using `IExecuteFunctions` as the context object in AI nodes.
```
Comment (OlegIvaniv) [inline packages/core/src/NodeExecuteFunctions.ts:4049]: This comment is no longer relevant, right?

File: packages/core/src/NodeExecuteFunctions.ts [hunk 2]
```diff
@@ -4059,6 +4008,258 @@ export function getExecuteFunctions(
...
+			return NodeHelpers.getNodeOutputs(workflow, node, nodeType.description).map((output) => {
+				if (typeof output === 'string') {
+					return {
+						type: output,
+					};
+				}
+				return output;
+			});
+		},
+		async getInputConnectionData(inputName, itemIndex): Promise<unknown> {
```
Comment (OlegIvaniv) [inline packages/core/src/NodeExecuteFunctions.ts:4085]:
```suggestion
		async getInputConnectionData(inputName: NodeConnectionType, itemIndex: number): Promise<unknown> {
```

File: packages/core/src/NodeExecuteFunctions.ts [hunk 3]
```diff
@@ -4059,6 +4008,258 @@ export function getExecuteFunctions(
...
+				node,
+				parameterName,
+				itemIndex,
+				mode,
+				getAdditionalKeys(additionalData, mode, runExecutionData),
+				executeData,
+				fallbackValue,
+				options,
+			)) as ISupplyDataFunctions['getNodeParameter'],
+		getWorkflowDataProxy: (itemIndex) =>
```
Comment (OlegIvaniv) [inline packages/core/src/NodeExecuteFunctions.ts:4143]:
```suggestion
		getWorkflowDataProxy: (itemIndex: number) =>
```

File: packages/core/src/NodeExecuteFunctions.ts [hunk 4]
```diff
@@ -4059,6 +4008,258 @@ export function getExecuteFunctions(
...
+					additionalData.sendDataToUI('sendConsoleMessage', {
+						source: `[Node: "${node.name}"]`,
+						messages: args,
+					});
+				}
+			} catch (error) {
+				Logger.warn(`There was a problem sending message to UI: ${error.message}`);
+			}
+		},
+		logAiEvent: async (eventName, msg: string) =>
```
Comment (OlegIvaniv) [inline packages/core/src/NodeExecuteFunctions.ts:4182]:
```suggestion
		logAiEvent: async (eventName: AiEvent, msg: string) =>
```

File: packages/core/src/NodeExecuteFunctions.ts [hunk 5]
```diff
@@ -4059,6 +4008,258 @@ export function getExecuteFunctions(
...
+			additionalData.logAiEvent(eventName, {
+				executionId: additionalData.executionId ?? 'unsaved-execution',
+				nodeName: node.name,
+				workflowName: workflow.name ?? 'Unnamed workflow',
+				nodeType: node.type,
+				workflowId: workflow.id ?? 'unsaved-workflow',
+				msg,
+			}),
+		addInputData(connectionType, data): { index: number } {
```
Comment (OlegIvaniv) [inline packages/core/src/NodeExecuteFunctions.ts:4191]:
```suggestion
		addInputData(connectionType: NodeConnectionType, data: INodeExecutionData[][]): { index: number } {
```

File: packages/core/src/NodeExecuteFunctions.ts [hunk 6]
```diff
@@ -4059,6 +4008,258 @@ export function getExecuteFunctions(
...
+				Logger.warn(
+					`There was a problem logging input data of node "${this.getNode().name}": ${
+						error.message
+					}`,
+				);
+			});
+
+			return { index: currentNodeRunIndex };
+		},
+		addOutputData(connectionType, currentNodeRunIndex, data): void {
```
Comment (OlegIvaniv) [inline packages/core/src/NodeExecuteFunctions.ts:4218]:
```suggestion
		addOutputData(connectionType: NodeConnectionType, currentNodeRunIndex: number, data: INodeExecutionData[][]): void {
```

**Review by OlegIvaniv — COMMENTED** [review]
Thanks for this refactor! The AI nodes and wrapped node tools seems to be working as expected. Added just a few suggestions to improve types

File: packages/core/src/NodeExecuteFunctions.ts [hunk 7]
```diff
@@ -4059,6 +4008,258 @@ export function getExecuteFunctions(
...
+				itemIndex,
+				// TODO: revert this back to `node.name` when we stop using `IExecuteFunctions` as the context object in AI nodes.
+				// https://linear.app/n8n/issue/CAT-269
+				node.name,
+				connectionInputData,
+				mode,
+				getAdditionalKeys(additionalData, mode, runExecutionData),
+				executeData,
+			),
+		executeWorkflow: async (workflowInfo, inputData, parentCallbackManager) =>
```
Comment (OlegIvaniv) [inline packages/core/src/NodeExecuteFunctions.ts:4057]:
```suggestion
		executeWorkflow: async (
		workflowInfo: IExecuteWorkflowInfo,
		inputData?: INodeExecutionData[],
		parentCallbackManager?: CallbackManager) =>
```

**Review by OlegIvaniv — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.66.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.66.0)

**Pull request merged**